src/jit/flowgraph.cpp

```diff
@@ -23827,3 +23827,237 @@ void Compiler::fgCleanupContinuation(BasicBlock* continuation)
     assert(foundEndLFin);
 #endif // !FEATURE_EH_FUNCLETS
 }
+
+class FatCalliTransoformer
+{
+private:
+    Compiler* compiler;
+
+    FatCalliTransoformer(Compiler* compiler) : compiler(compiler)
+    {
+    }
+
+    void Run()
+    {
+        BasicBlock* firstBasicBlock = compiler->fgFirstBB;
+        for (BasicBlock* block = firstBasicBlock; block != nullptr; block = block->bbNext)
+        {
+            TransformBlock(block);
+        }
+    }
+
+    bool ContainsFatCalli(GenTreeStmt* stmt)
+    {
+        GenTreePtr fatPointerCandidate = stmt->gtStmtExpr;
+        if (fatPointerCandidate->OperIsAssignment())
+        {
+            fatPointerCandidate = fatPointerCandidate->gtGetOp2();
+        }
+        return fatPointerCandidate->IsCall() && fatPointerCandidate->gtFlags & GTF_CALL_M_FAT_POINTER_CHECK;
+    }
+
+    class StatementTransformer
+    {
+    private:
+        Compiler*    compiler;
+        BasicBlock*  currBlock;
+        BasicBlock*  remainderBlock;
+        BasicBlock*  checkBlock;
+        BasicBlock*  thenBlock;
+        BasicBlock*  elseBlock;
+        GenTreeStmt* stmt;
+        GenTreePtr   call;
+        GenTreePtr   fptrAddress;
+        var_types    pointerType;
+        bool         doesReturnValue;
+
+        const int FAT_POINTER_MASK = 0b00000010;
+        const int HIGH_PROBABILITY = 80;
+
+        StatementTransformer(Compiler* compiler, BasicBlock* block, GenTreeStmt* stmt)
+            : compiler(compiler), currBlock(block), stmt(stmt)
+        {
+            remainderBlock = checkBlock = thenBlock = elseBlock = nullptr;
+            doesReturnValue                                     = stmt->gtStmtExpr->OperIsAssignment();
+            SetFptrAddress();
+            pointerType = fptrAddress->TypeGet(); // FIXME
+        }
+
+        BasicBlock* CreateAndInsertBasicBlock(BBjumpKinds jumpKind, BasicBlock* insertAfter)
+        {
+            BasicBlock* block = compiler->fgNewBBafter(jumpKind, insertAfter, true);
+            if ((insertAfter->bbFlags & BBF_INTERNAL) == 0)
+            {
+                block->bbFlags &= ~BBF_INTERNAL;
+                block->bbFlags |= BBF_IMPORTED;
+            }
+            return block;
+        }
+
+        void CreateRemainder()
+        {
+            remainderBlock = compiler->fgSplitBlockAfterStatement(currBlock, stmt);
```

Comment:
Can't this transform be done right after `fgRemovePreds` is called instead of just before it?